OpenRA.Mods.Common/Activities/DonateCash.cs

```diff
@@ -21,16 +21,52 @@ class DonateCash : Enter
 		readonly int payload;
 		readonly int playerExperience;
 
+		Actor enterActor;
+
 		public DonateCash(Actor self, in Target target, int payload, int playerExperience, Color? targetLineColor)
 			: base(self, target, targetLineColor)
 		{
 			this.payload = payload;
 			this.playerExperience = playerExperience;
 		}
 
+		protected override bool TryStartEnter(Actor self, Actor targetActor)
+		{
+			enterActor = targetActor;
+
+			// Make sure the target actor is still owned by a player with eligible relationship
+			var isTargetStillValid = true;
+			var targetInfo = targetActor.Info.TraitInfoOrDefault<AcceptsDeliveredCashInfo>();
+			if (targetInfo == null
+				|| !targetInfo.ValidRelationships.HasRelationship(targetActor.Owner.RelationshipWith(self.Owner))
+				|| (targetInfo.SamePlayerOnly && targetActor.Owner != self.Owner))
+				isTargetStillValid = false;
```

Comment:
Could we extract this to a new method called `ValidTarget` or something and use it here and below?